Add a show_leading_ns option

kevans91 opened the pull request:
This is used in scenarios where a user may not have full access to a namespace,
but they might have access to pages within a namespace. For instance, consider
the following layout:

FAQ
- Restricted
- - Entry1
- - Entry2
- - Entry3
- Public
- - Entry1
- - Entry2
- - Entry3

Along with the following ACLs:

FAQ:Public:* -> @ALL/Read
FAQ:Restricted:* -> @RestrictedPeople/read, @ALL/None
FAQ:Restricted:Entry3 -> @FurtherRestrictedPeople/read

Such a setup allows RestrictedPeople complete access to the Restricted namespace,
while giving FurtherRestrictedPeople limited insight into the namespace -- as far
as they can read into it.

This option was designed for interactions w/ hide_acl_nsnotr and show_acl properly.
With show_acl, still only FAQ:Restricted:Entry3 would be shown if a user didn't have
ACLs for FAQ:Restricted.

Commit:
Add a show_leading_ns option

This is used in scenarios where a user may not have full access to a namespace,
but they might have access to pages within a namespace. For instance, consider
the following layout:

FAQ
- Restricted
- - Entry1
- - Entry2
- - Entry3
- Public
- - Entry1
- - Entry2
- - Entry3

Along with the following ACLs:

FAQ:Public:* -> @ALL/Read
FAQ:Restricted:* -> @RestrictedPeople/read, @ALL/None
FAQ:Restricted:Entry3 -> @FurtherRestrictedPeople/read

Such a setup allows RestrictedPeople complete access to the Restricted namespace,
while giving FurtherRestrictedPeople limited insight into the namespace -- as far
as they can read into it.

This option was designed for interactions w/ hide_acl_nsnotr and show_acl properly.
With show_acl, still only FAQ:Restricted:Entry3 would be shown if a user didn't have
ACLs for FAQ:Restricted.

Commit:
show_leading_ns: Fix interaction w/ show_acl and hide_acl_nsnotr

Using a slightly altered version of our previous example:

FAQ
- Restricted
- - Entry1
- - Entry2
- - Entry3
- Other Restricted
- - Entry1
- - Entry2
- - Entry3

One with no read permissions in the entire FAQ tree beyond
FAQ:Restricted:Entry3 would have the FAQ:Other_Restricted namespace exposed
to them in some combinations of show_acl and hide_acl_nsnotr. This is
technically incorrect, as said user only has limited exposure to these
namespaces.

Fix the semantics by ensuring that we don't enumerate other nested
namespaces if we're inside a namespace that we have limited access to.
Limited access is only checked if show_leading_ns is enabled, otherwise we
assume that we're OK.

xif-fr commented:
Thanks for the contribution. Looks good. I'll merge it when I would have tested it.

Commit:
Convert two more ACL checks to using _cached_quickaclcheck

ACL checks can be prohibitively expensive; these ones killed our performance as
we added more ACLs. Converting these to also using _cached_quickaclcheck (which
caches the results of auth_quickaclcheck and should be harmless for this
purposes) boosts our performance back up to expected levels.

kevans91 commented:
Thanks! I converted a couple more auth_quiickaclcheck to _cached_quickaclcheck calls because these lookups were becoming more and more expensive as we added ACLs.

xif-fr commented:
I changed a bit the semantics with respect to show_acl. Here it is, for namespaces :
![catlist-ns-perms](https://user-images.githubusercontent.com/34941891/55121955-da867d00-50f4-11e9-8b0b-ba3c21afe1e7.png)
When I say 'recurse', only items with read permission are show by default.
